content/blog/2022-04-04-develop-and-test-data-analytics-pipelines-on-localstack/index.md

```diff
@@ -0,0 +1,236 @@
+
+```python
+# data_pipeline_stack.py
+
+
+kinesis_stream = aws_kinesis.Stream(
+    self,
+    "data_pipeline_kinesis_stream",
+    shard_count=1,
+)
+
+monitored_log_group = aws_logs.LogGroup.from_log_group_arn(
+    self, "monitored_log_group", monitored_log_group_arn
+)
+monitored_log_group.add_subscription_filter(
+    "monitored_log_group_cloudwatch_subscription",
+    destination=aws_logs_destinations.KinesisDestination(kinesis_stream),
+    filter_pattern=aws_logs.FilterPattern.all(
+        aws_logs.FilterPattern.exists("$.event_class"),
+        aws_logs.FilterPattern.exists("$.event_type"),
+    ),
+)
+
+```
+First we stand up a Kinesis stream for our event data. Next, we import the log group associated with the Lambda function we're interested in using its ARN. Finally, we declare a subscription filter using these two elements. The subscription filter acts as the "glue" between CloudWatch and Kinesis. Note the filter pattern: this ensures that only the structured events that we care about will be sent to Kinesis. We don't want every Lambda log message to show up in our data warehouse! Our filter only accepts log lines that meet the following criteria:
+* JSON encoded
+* Contains an `event_class` field
+* Contains an `event_type` field
+
+This filter was chosen based on the particular analytics event format we use at LocalStack.
+Finally, let's deploy a "loader Lambda" function to consume events from Kinesis and load them into Tinybird. You can find the function code [here](https://github.com/localstack/support-bot-analytics/blob/simplified-blog/lambdas/kinesis_tinybird_loader/kinesis_tinybird_loader.py)
+
+```python
+# data_pipeline_stack.py
+
+
+loader_lambda = aws_lambda.Function(
+    self,
+    "kinesis_tinybird_loader",
+    runtime=aws_lambda.Runtime.PYTHON_3_9,
+    handler="kinesis_tinybird_loader.handler",
+    code=aws_lambda.Code.from_asset(
+        path.join(
+            path.realpath(path.dirname(__file__)),
+            "../..",
+            "kinesis_tinybird_loader_pkg.zip",
+        )
+    ),
+    timeout=Duration.seconds(7),
+    environment={
+        "TINYBIRD_AUTH_TOKEN": tinybird_auth_token,
+        "TINYBIRD_URL": tinybird_url,
+    },
+)
+loader_lambda.add_event_source(
+    aws_lambda_event_sources.KinesisEventSource(
+        kinesis_stream,
+        starting_position=aws_lambda.StartingPosition.LATEST,
+        retry_attempts=20,
+        batch_size=200,
+        max_batching_window=Duration.minutes(2),
+    )
+)
+```
+
+## Mocking External Resources
+We've defined the data pipeline, but in order to test it end-to-end locally we need to mock two additional components. We need to set up a Lambda function with a log group that we can monitor for events, plus a local endpoint that can imitate the Tinybird API.
+
+### Logger Lambda for Testing
+Since Cloudwatch logging is integrated with Lambda out of the box the code for emitting a structured analytics event as a log message is dead simple:
```

Comment:
Building on @whummer suggestion:

```suggestion
Since CloudWatch logging is integrated with Lambda out of the box, the code for emitting a structured analytics event as a log message is dead simple:
```